northstar/src/runtime/repository.rs

```diff
@@ -85,38 +89,46 @@ impl DirRepository {
 
         let start = Instant::now();
         let mut loads = vec![];
-        let npk_extension = Some(OsStr::new("npk"));
+        let blacklist = Arc::new(blacklist.clone());
         while let Ok(Some(entry)) = readir.next_entry().await {
             let file = entry.path();
-            if file.extension() == npk_extension {
-                let task = task::spawn_blocking(move || {
-                    debug!(
-                        "Loading {}{}",
-                        file.display(),
-                        if key.is_some() { " [verified]" } else { "" }
-                    );
-                    let npk = Npk::from_path(&file, key.as_ref())
-                        .map_err(|e| Error::Npk(file.display().to_string(), e))?;
-                    let name = npk.manifest().name.clone();
-                    let version = npk.manifest().version.clone();
-                    let container = Container::new(name, version);
-                    Result::<_, Error>::Ok((container, file, npk))
-                })
-                .then(|r| match r {
-                    Ok(r) => ready(r),
-                    Err(_) => panic!("Task error"),
-                });
-                loads.push(task);
-            } else {
-                debug!("Skipping {}", file.display());
-            }
+            let blacklist = blacklist.clone();
+            let task = task::spawn_blocking(move || {
+                debug!(
+                    "Loading {}{}",
+                    file.display(),
+                    if key.is_some() { " [verified]" } else { "" }
+                );
+                let npk = Npk::from_path(&file, key.as_ref())
+                    .map_err(|e| Error::Npk(file.display().to_string(), e))?;
+                let name = npk.manifest().name.clone();
+                let version = npk.manifest().version.clone();
+                let container = Container::new(name, version);
+
+                if blacklist.contains(&container) {
+                    Err(Error::DuplicateContainer(container))
+                } else {
+                    Ok((container, file, npk))
+                }
+            })
+            .then(|r| match r {
+                Ok(r) => ready(r),
+                Err(_) => panic!("Task error"),
+            });
+            loads.push(task);
         }
 
         let results = join_all(loads).await;
         for result in results {
             match result {
                 Ok((container, file, npk)) => {
-                    containers.insert(container, (file, Arc::new(npk)));
+                    // If the container name/version is already in there remove the present
+                    // container and skip the newly parsed one.
+                    if containers.remove(&container).is_some() {
+                        warn!("Skipping duplicate container {}", container);
```

Comment:
Isn't the presence of a duplicate container a fatal error and we should fail?

Reply:
I thought about that, but it can leave Northstar in a situation where it's not able to start anymore. Think we had the "be pedantic or not" topic somewhere else.

Reply:
That was the autostart behaviour which ended up being configurable. Not sure what the best way to go is here.

Reply:
Yes. Right. Regarding the autostart thing it was easy to push the decision to the user.